core/ledger/kvledger/snapshot_mgmt.go

```diff
@@ -0,0 +1,329 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package kvledger
+
+import (
+	"math"
+	"os"
+
+	"github.com/hyperledger/fabric/common/ledger/util"
+	"github.com/hyperledger/fabric/common/ledger/util/leveldbhelper"
+	"github.com/pkg/errors"
+)
+
+var (
+	snapshotRequestKeyPrefix        = []byte("s")
+	defaultSmallestHeight    uint64 = math.MaxUint64
+)
+
+// SubmitSnapshotRequest submits a snapshot request for the specified height.
+// The request will be stored in the ledger until the ledger's block height is equal to
+// the specified height and the snapshot generation is completed.
+// When height is 0, it will generate a snapshot at the current block height.
+// It returns an error if the specified height is smaller than the ledger's block height.
+// It uses snapshotRequestLock to synchronize among concurrent SubmitSnapshotRequest
+// and CancelSnapshotRequest calls and uses snapshotGenerationRWLock to synchronize with commit.
+// It starts a go-routine to call generateSnapshot if needed and will keep the snapshotGenerationRWLock
+// until generateSnapshot is finished.
+func (l *kvLedger) SubmitSnapshotRequest(height uint64) error {
+	l.snapshotRequestLock.Lock()
+	defer l.snapshotRequestLock.Unlock()
+
+	// before calling GetBlockchainInfo, locks snapshotGenerationRWLock to synchronize with commit.
+	// It gets the block height and invokes generateSnapshot if the submitted height matches
+	// the block height. The snapshotGenerationRWLock will be held and unlocked inside of generateSnapshot.
+	keepLock := false
+	l.snapshotGenerationRWLock.RLock()
```

Comment:
Would it make sense to use `blockCommit.RLock()` instead of `snapshotGeneration.RLock()`?